api/group_creation/group_manager.py

```diff
@@ -0,0 +1,150 @@
+                    saturday=group_candidate.saturday,
+                    sunday=group_candidate.sunday,
+                )
+
+                group_creation_timestamp = timezone.now()
+                StatusSetter.set_status(
+                    obj=group, status=GroupStatus.PENDING, status_since=group_creation_timestamp
+                )
+                group.save()
+                group.teachers.add(group_candidate.teacher)
+                group.students.set(group_candidate.students)
+
+                # QQ: do we update teacher/student status automatically,
+                # or wait for coordinator action?
+                next_teacher_status = (
+                    TeacherStatus.TEACHING_ANOTHER_GROUP_OFFERED
+                    if group_candidate.teacher.status == TeacherStatus.TEACHING_ACCEPTING_MORE
+                    else TeacherStatus.GROUP_OFFERED
+                )
+                StatusSetter.set_status(
+                    obj=group_candidate.teacher,
+                    status=next_teacher_status,
+                    status_since=group_creation_timestamp,
+                )
+                GroupManager._create_log_events(group)
+                return group
+        # No suitable groups found
+        return None
+
+    @staticmethod
+    def _get_allowed_group_size(age_range: AgeRange) -> GroupSizeRestriction:
+        if MAX_AGE_TEEN_GROUP < age_range.age_from <= age_range.age_to:
+            return GroupManager.GroupSizeRestriction(min_number=5, max_number=10)
+        if MAX_AGE_KIDS_GROUP <= age_range.age_from <= age_range.age_to <= MAX_AGE_TEEN_GROUP:
+            return GroupManager.GroupSizeRestriction(min_number=2, max_number=8)
+        if 0 <= age_range.age_from <= age_range.age_to <= MAX_AGE_KIDS_GROUP:
+            return GroupManager.GroupSizeRestriction(min_number=2, max_number=6)
+        # If age range does not fall into expected ranges, fail early
+        raise ValueError(f"Group age range is inconsistent with boundaries: {age_range}")
+
+    @staticmethod
+    def _iterate_group_candidates(teacher_id: int) -> Iterable[GroupCandidate]:
+        # THIS IS A STUB METHOD RETURNING DUMMY VALUES
+        # TODO filter for language+level, communication language, age groups, time slots (all UTC?)
+        # TODO iterate in correct priority order, yield results
+        teacher = Teacher.objects.filter(personal_info__id=teacher_id).get()
+
+        yield GroupManager.GroupCandidate(
+            age_range=AgeRange(age_from=18, age_to=90),
+            language_and_level=teacher.teaching_languages_and_levels.first(),  # type: ignore
+            communication_language_mode=CommunicationLanguageMode(
+                teacher.personal_info.communication_language_mode
+            ),
+            teacher=teacher,
+            students=list(range(1, 10)),  # type: ignore
+            monday=timezone.now(),
+            wednesday=timezone.now(),
+        )
+
+    # QQ: do we need this?
```

Comment:
Yes, we do need to create log events. BTW please have a look at `models/log_event.py`. I actually think that it was wrong of me to put `on_delete=models.CASCADE` for `.group` in all subclasses of `LogEvent` except `GroupLogEvent`. This is because we will most likely be deleting tentative groups as I don't see any reason to keep thousands of auto-created tentative groups in the database. The log events, however, I would want to keep, to see how many times a student and a teacher refused to take on a group. It should definitely remain `CASCADE` in GroupLogEvent, though.

Reply:
I'm not sure I follow. So, currently all log events are deleted as CASCADE but it sounds reasonable to me. If we delete group-related log events, won't log events for students and teachers stay in place for analytics still? We wll delete a lot of groups, yes, but we'll keep students and teachers.

Reply:
Maybe I misunderstand the mechanic of `CASCADE`, but I think that if you delete a `Group` then, in case of `on_delete=CASCADE`, all log events that were related to this group will also be deleted. And student- and teacher-related log events have `.to_group` and `.from_group` attributes, linking these log events to `Group`. So I'm guessing that these log events will be deleted as well if a tentative `Group` is deleted.

Reply:
Oh yes! I missed there were references to groups in these events too. Then, would setting to `None` instead work for us? Actually I don't see other ways, but to confirm.

Also I believe we should manually delete group-related log events, so they don't hang around as they are meaningless without the group? I'll do that now